File: matrix_block.py
import numpy as np

def multiply_block(A, B, block_size=64):
    """
    Blocked (cache-friendly) matrix multiplication.
    Splits A and B into sub-blocks to improve cache locality.
    Works for any square size n (not necessarily a multiple of block_size).
    """
    n = A.shape[0]
    C = np.zeros((n, n), dtype=A.dtype)

    for i in range(0, n, block_size):
        i_end = min(i + block_size, n)
        for k in range(0, n, block_size):
            k_end = min(k + block_size, n)
            # Preload the A sub-block to reduce repeated slicing
            A_block = A[i:i_end, k:k_end]
            for j in range(0, n, block_size):
                j_end = min(j + block_size, n)
                # Multiply sub-blocks and accumulate
                C[i:i_end, j:j_end] += A_block @ B[k:k_end, j:j_end]
    return C

File: java_code/MatrixBlock.java
public class MatrixBlock {

    // Blocked (cache-friendly) matrix multiplication
    public static double[][] multiplyBlock(double[][] A, double[][] B, int blockSize) {
        int n = A.length;
        double[][] C = new double[n][n];

        for (int i = 0; i < n; i += blockSize) {
            for (int j = 0; j < n; j += blockSize) {
                for (int k = 0; k < n; k += blockSize) {

                    int iEnd = Math.min(i + blockSize, n);
                    int jEnd = Math.min(j + blockSize, n);
                    int kEnd = Math.min(k + blockSize, n);

                    for (int ii = i; ii < iEnd; ii++) {
                        for (int jj = j; jj < jEnd; jj++) {
                            double sum = 0;
                            for (int kk = k; kk < kEnd; kk++) {
                                sum += A[ii][kk] * B[kk][jj];
                            }
                            C[ii][jj] += sum;
                        }
                    }

                }
            }
        }

        return C;
    }

    // Small test
    public static void main(String[] args) {

        double[][] A = MatrixBasic.randomMatrix(8);
        double[][] B = MatrixBasic.randomMatrix(8);

        double[][] C_block = multiplyBlock(A, B, 4);
        double[][] C_basic = MatrixBasic.multiplyBasic(A, B);

        boolean ok = true;

        for (int i = 0; i < C_block.length; i++) {
            for (int j = 0; j < C_block.length; j++) {
                if (Math.abs(C_block[i][j] - C_basic[i][j]) > 1e-6) {
                    ok = false;
                    break;
                }
            }
        }

        System.out.println("Block correct: " + ok);
    }
}
